sklearn/utils/__init__.py

```diff
@@ -178,22 +178,66 @@ def axis0_safe_slice(X, mask, len_mask):
     return np.zeros(shape=(0, X.shape[1]))
 
 
-def safe_indexing(X, indices):
+def safe_indexing(X, indices, axis=0):
+    """Return rows, items or columns of X using indices.
+
+    Parameters
+    ----------
+    X : array-like, sparse-matrix, list, pandas.DataFrame, pandas.Series
+        Data from which to sample rows, items or columns.
+    indices : array-like
+        - When ``axis=0``, indices need to be an array of integer.
+        - When ``axis=1``, indices can be one of:
```

Comment:
I find this description a bit confusing.

What's the difference between a key type and a key data type? Is it really worth separating them into 2 different sublists?

why is integer treated differently from scalar?

Reply:
This is from the original docstring. I assume that data type is the type of data when you are using a container (slice, array, list, etc.)